### PR TITLE
Emit error in logs if keytab files can't be opened

### LINKED ISSUE #228: When the GssapiCredStore file is absent, the log message is not very clear
I made a typo when setting `GssapiCredStore`, and the log message was `SPNEGO cannot find mechanisms to
negotiate`, which does not point admins in the right direction.

### PR DESCRIPTION
**Pull request by simo5**
This will give a useful warning to admins when config point to missing
files.

Fixes #228

### CODE REVIEW

File: src/mod_auth_gssapi.c [hunk 1]
```diff
@@ -1532,6 +1532,23 @@ static const char *mag_cred_store(cmd_parms *parms, void *mconfig,
     }
     cfg->cred_store->count++;
 
+    /* check for files that we know should be present, so admins get
+     * some rope to figure out issues when they cannot be accessed */
+    if ((strcmp(key, "keytab") == 0) ||
+        (strcmp(key, "client_keytab") == 0)) {
```
Comment (frozencemetery) [inline src/mod_auth_gssapi.c:1538]: Extra parens aren't needed around the two conditions

File: src/mod_auth_gssapi.c [hunk 2]
```diff
@@ -1532,6 +1532,23 @@ static const char *mag_cred_store(cmd_parms *parms, void *mconfig,
...
+        apr_status_t rc;
+        apr_file_t *file;
+        rc = apr_file_open(&file, value, APR_FOPEN_READ, 0, parms->pool);
+        if (rc != APR_SUCCESS) {
+            char err[256];
+            apr_strerror(rc, err, sizeof(err));
+            ap_log_error(APLOG_MARK, APLOG_ERR, 0, parms->server,
+                    "Cannot open %s file %s: %s", key, value, err);
```
Comment (frozencemetery) [inline src/mod_auth_gssapi.c:1546]: Indentation here seems wonky - shouldn't this line up to the `APLOG_MARK`?

File: src/mod_auth_gssapi.c [hunk 3]
```diff
@@ -1532,6 +1532,23 @@ static const char *mag_cred_store(cmd_parms *parms, void *mconfig,
...
+        } else {
```
Comment (frozencemetery) [inline src/mod_auth_gssapi.c:1547]: Shouldn't this error cause a function exit, like we do above?  (That would also let you remove this `else`-block.)

File: tests/magtests.py [hunk 4]
```diff
@@ -443,7 +444,9 @@ def setup_http(testdir, so_dir, wrapenv):
     })
 
     httpd = "httpd" if distro == "Fedora" else "apache2"
+    log = open(httpdstdlog, 'a')
     httpproc = subprocess.Popen([httpd, '-DFOREGROUND', '-f', config],
+                                stdout=log, stderr=log,
```
Comment (frozencemetery) [inline tests/magtests.py:449]: I think the preferred idiom for this is `stderr=subprocess.STDOUT`?

File: tests/magtests.py [hunk 5]
```diff
@@ -803,6 +808,22 @@ def test_mech_name(testdir, testenv, logfile):
     return 0
 
 
+def test_file_check(testdir, testenv, logfile):
+    basicdir = os.path.join(testdir, 'httpd', 'html', 'keytab_file_check')
+    os.mkdir(basicdir)
+    shutil.copy('tests/index.html', basicdir)
+
+    filec = subprocess.Popen(["tests/t_file_check.py"],
+                             stdout=logfile, stderr=logfile,
```
Comment (frozencemetery) [inline tests/magtests.py:817]: (As above)

File: tests/t_file_check.py [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+import requests
+from requests.auth import HTTPBasicAuth
+
+
+if __name__ == '__main__':
+    url = 'http://%s/keytab_file_check/' % os.environ['NSS_WRAPPER_HOSTNAME']
+    r = requests.get(url, auth=HTTPBasicAuth(os.environ['MAG_USER_NAME'],
+                                             os.environ['MAG_USER_PASSWORD']))
+    if r.status_code != 200:
+        raise ValueError('Basic Auth Failed(Keytab File Check)')
```
Comment (frozencemetery) [inline tests/t_file_check.py:15]: Maybe missing space?

I guess we don't really have a style for these.  Some we've used are

```python
raise ValueError('Basic Auth: Failed Authentication')
```

but we also have things like

```python
raise ValueError('Spnego Negotiate Once failed - 401 expected')
```

and

```python
raise ValueError('Localname, REMOTE_USER check failed (%s != %s)' % (
                 line, os.environ['MAG_REMOTE_USER']))
```

As long as it's readable it of course doesn't really matter :)

**Review by frozencemetery — COMMENTED** [review]
Substantively good; some minor things inline.

File: src/mod_auth_gssapi.c (hunk 3 above)
Comment (simo5) [reply src/mod_auth_gssapi.c:1547]: no
I want the thing not to crash the apache server
it is possible during deployment people add a client _ketab later
I just want admins to see the error if something doesn't work, not a hard fail

File: src/mod_auth_gssapi.c (hunk 1 above)
Comment (simo5) [reply src/mod_auth_gssapi.c:1538]: ok

File: src/mod_auth_gssapi.c (hunk 2 above)
Comment (simo5) [reply src/mod_auth_gssapi.c:1546]: yeah

File: tests/magtests.py (hunk 4 above)
Comment (simo5) [reply tests/magtests.py:449]: I think this way is more clear that I am redirecting both outputs to the same logfile

File: tests/t_file_check.py (hunk 6 above)
Comment (simo5) [reply tests/t_file_check.py:15]: heh it's just a string to match in some debugging session, it's fine :)

**Review by frozencemetery — APPROVED** [review]

**Pull request merged**